File: Code/C++/Postfixexpression.cpp
//Evaluation of Postfix expression

/*A postfix expression is a combination of operands and operators in which the operators are placed after the operands.
In the Evaluation of the postfix expression, we are given a string which contains postfix expression.We have to evaluate that expression.
The main idea of the evaluation of postfix expression is as follows:
While reading the expression from left to right we should push the element in the stack if it is an operand.
Pop the two operands from the stack, if the element is an operator and then evaluate it.
Push back the result of the evaluation. Repeat it till the end of the expression.
*/

#include <iostream>
#include <math.h>
using namespace std;

class stack {
  int top;
  float stk[50];
  public:
    stack() {
      top = -1;
    }
  void push(float p) {
    stk[++top] = p;
  }
  float pop() {
    return stk[top--];
  }
};

int main()

{
  char p[50];
  float data[50], a, b;
  stack st;
  cout << "Enter Postfix Expression with out any blank spaces:";
  cin >> p;
  int i = 0;
  while (p[i] != '\0') {
    if (isalpha(p[i])) {
      cout << "Enter values of" << p[i] << ":";
      cin >> data[i];
    }
    i++;
  }
  p[i] = ')';
  i = 0;
  while (p[i] != ')') {
    switch (p[i]) {
    case '+':
      a = st.pop();
      b = st.pop();

      st.push(b + a);
      break;
    case '-':
      a = st.pop();
      b = st.pop();
      st.push(b - a);
      break;
    case '*':
      a = st.pop();
      b = st.pop();
      st.push(b * a);
      break;
    case '/':
      a = st.pop();
      b = st.pop();
      st.push(b / a);
      break;
    case '^':
      a = st.pop();
      b = st.pop();
      st.push(pow(b, a));
      break;
    default:
      st.push(data[i]);
    }
    i++;
  }
  p[i] = '\0';
  cout << "The given postfix expression is:";
  for (i = 0; p[i]; i++)

  {
    if (isalpha(p[i]))
      cout << data[i] << " ";
    else
      cout << p[i] << " ";
  }
  cout << "\nThe result of evaluated postfix expression is:" << st.pop();
}

/* Test cases:
1.
Input:

Enter Postfix Expression with out any blank spaces:abc*+                                                              
Enter values ofa:1                                                                                                    
Enter values ofb:2                                                                                                    
Enter values ofc:3

Output:

The given postfix expression is:1 2 3 * +                                                                             
The result of evaluated postfix expression is:7 

2.
Input:

Enter Postfix Expression with out any blank spaces:abcd*-+                                                            
Enter values ofa:4                                                                                                    
Enter values ofb:3                                                                                                    
Enter values ofc:2                                                                                                    
Enter values ofd:1 

Output:

The given postfix expression is:4 3 2 1 * - +  
The result of evaluated postfix expression is:5 

Time Complexity : O(n) where 'n' is the size of the given string.
Space Complexity: O(n) because we used the stack space for 'n' characters.
*/

File: Code/Python/Postfix_expression.py
'''
The Postfix notation is used to represent algebraic expressions.
Using stack, we can easily evaluate postfix expression there will be only two scenarios. We have to scan string from left to right. If we encounter operator while we are scanning the string then we will pop two elements from the stack and perform the operation of current operator and we will push back the result into the stack. And if we encounter operand just push that operand into the stack.
Steps for evaluating postfix expression.

1. Accept postfix expression string in post variable.
2.
For i in post:
    If i is an operand:
        Push i in stack.
    Else:
        Pop two elements from stack e.g. X and Y
        Perform operation with current operator on both the parents i.e X i Y
        Push the result back into the stack.
End for loop.
3. Finally pop out the result from the stack.


'''
# Python program to evaluate value of a postfix expression

class evaluate_postfix:
    def __init__(self):
        self.items=[]
        self.size=-1
    def isEmpty(self):
        return self.items==[]
    def push(self,item):
        self.items.append(item)
        self.size+=1
    def pop(self):
        if self.isEmpty():
            return 0
        else:
            self.size-=1
            return self.items.pop()
    def seek(self):
        if self.isEmpty():
            return False
        else:
            return self.items[self.size]
    def evalute(self,expr):
        for i in expr:
            if i in '0123456789':
                self.push(i)
            else:
                op1=self.pop()
                op2=self.pop()
                result=self.cal(op2,op1,i)
                self.push(result)
        return self.pop()
    def cal(self,op2,op1,i):
        if i == '*':
            return int(op2)*int(op1)
        elif i == '/':
            return int(op2)/int(op1)
        elif i == '+':
            return int(op2)+int(op1)
        elif i == '-':
            return int(op2)-int(op1)
s=evaluate_postfix()
expr=input('enter the postfix expression : ')
value=s.evalute(expr)
print('The result of postfix expression',expr,'is',value)

'''
Testcases:
Input: enter the prefix expression : 231*+9-
Output: The result of postfix expression 231*+9- is -4


Input: enter the prefix expression : ++596^-64
Output: The result of postfix expression ++596^-64 is 20
Time complexity: O(n) where n is number of characters in input expression.
Space Complexity: O(n) We used the stack space for n characters.

'''